Swarm mode reports invalid hostname if hostname is changed after Docker daemon is started

**Description**

If the hostname of a machine is changed before Docker is started and after swarm mode is enabled, Swarm mode seems to report the old (invalid) hostname.

**Steps to reproduce the issue:**

You dupe this fairly simply with `docker-machine` and two node cluster (probably even with one really).

First `create`.  Machine sets host name according to machine name automatically.  And Docker is started on these machines of course.

Then change host name on created machines.

Then start swarm init / join.

Here's a little script you can use to dupe if you want.  You should see that `hostname` command will report the correct set hostname when executed after the middle steps to set it FWIW.

``` console
docker-machine create -d virtualbox manager
docker-machine create -d virtualbox worker
docker-machine ssh worker 'sudo hostname baz && sudo sed -i "s/worker/baz/g" /etc/hosts && echo baz | sudo tee /etc/hostname'
docker-machine ssh manager 'sudo hostname quux && sudo sed -i "s/manager/quux/g" /etc/hosts && echo quux | sudo tee /etc/hostname'
docker-machine ssh manager docker swarm init --advertise-addr eth1
docker-machine ssh worker docker swarm join --token $(docker-machine ssh manager docker swarm join-token -q worker) $(docker-machine ip manager):2377
docker-machine ssh manager docker node ls
```

Cool, the hostname was set before swarm mode was initiated, so the new hostnames should be propagated when the swarm is started, right?

`docker node ls` reports the old values though.  These should say `quux` and `baz` for hostname.

```
$ docker-machine ssh manager docker node ls
ID                           HOSTNAME  STATUS  AVAILABILITY  MANAGER STATUS
3of2on83v5orfuozfqpnq00ul *  manager   Ready   Active        Leader
5usitwuan5g1n7l8drx15kue9    worker    Ready   Active
```

**Describe the results you received:**

Hostname is not reported correctly.

**Describe the results you expected:**

Hostname is reported correctly.

**Additional info**

If Docker daemon is restarted _before_ the Swarm initialization steps it seems to report the hostnames correctly.

**Output of `docker version`:**

```
Client:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.12.1
 API version:  1.24
 Go version:   go1.6.3
 Git commit:   23cf638
 Built:        Thu Aug 18 17:52:38 2016
 OS/Arch:      linux/amd64
```

(I seem to have duped this with `1.12.2-rc2` as well though)

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 0
Server Version: 1.12.1
Storage Driver: aufs
 Root Dir: /mnt/sda1/var/lib/docker/aufs
 Backing Filesystem: extfs
 Dirs: 0
 Dirperm1 Supported: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: overlay null bridge host
Swarm: active
 NodeID: 3of2on83v5orfuozfqpnq00ul
 Is Manager: true
 ClusterID: eqisp01jd77ahunfx80asdwez
 Managers: 1
 Nodes: 2
 Orchestration:
  Task History Retention Limit: 5
 Raft:
  Snapshot Interval: 10000
  Heartbeat Tick: 1
  Election Tick: 3
 Dispatcher:
  Heartbeat Period: 5 seconds
 CA Configuration:
  Expiry Duration: 3 months
 Node Address: 192.168.99.105
Runtimes: runc
Default Runtime: runc
Security Options: seccomp
Kernel Version: 4.4.17-boot2docker
Operating System: Boot2Docker 1.12.1 (TCL 7.2); HEAD : ef7d0b4 - Thu Aug 18 21:18:06 UTC 2016
OSType: linux
Architecture: x86_64
CPUs: 1
Total Memory: 995.9 MiB
Name: quux
ID: M7Q6:DORB:VZZB:WCJO:LVIG:247Y:BQCV:FZGH:VRMN:Y77H:XKA5:PECS
Docker Root Dir: /mnt/sda1/var/lib/docker
Debug Mode (client): false
Debug Mode (server): true
 File Descriptors: 32
 Goroutines: 124
 System Time: 2016-10-05T23:37:30.070842443Z
 EventsListeners: 0
Registry: https://index.docker.io/v1/
Labels:
 provider=virtualbox
Insecure Registries:
 127.0.0.0/8
```

**Cute animal photo**

Ladies and gentlemen, I present to you, the Hostname Corgi.

![image](https://cloud.githubusercontent.com/assets/1476820/19135573/6f8db3b2-8b1a-11e6-9b7f-b43164c6b7eb.png)

Thanks.

cc @aluzzardi @stevvooe @aaronlehmann FYI

I think I found the root cause. Getting errors in the logs that indicate that the node update isn't working, because an updated description is never found:
```bash
ERRO[0020] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0040] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0060] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0080] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0100] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0120] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
ERRO[0140] agent: updated node description unavailable   agent=&{client:0xc8202171a0} error=Get https://%2Fvar%2Frun%2Fdocker.sock/info: http: server gave HTTP response to HTTPS client
github.com/docker/swarmkit/vendor/github.com/docker/docker/client.(*Client).sendClientRequest
	/root/work/src/github.com/docker/swarmkit/vendor/github.com/docker/docker/client/request.go:143: error during connect module=node/agent
```
The error originates here
```go
desc, err := a.config.Executor.Describe(ctx)
```
The error indicates an HTTP response to HTTPS client. Figuring out why this is happening.

--- a/daemon/cluster/noderunner.go
+++ b/daemon/cluster/noderunner.go
@@ -94,8 +94,9 @@ func (n *nodeRunner) start(conf nodeStartConfig) error {
 		control = filepath.Join(n.cluster.runtimeRoot, controlSocket)
 	}
 
+	// Hostname is not set here. Instead, it is obtained from
+	// the node description that is reported periodically
 	swarmnodeConfig := swarmnode.Config{
-		Hostname:           n.cluster.config.Name,
 		ForceNewCluster:    conf.forceNewCluster,
 		ListenControlAPI:   control,
 		ListenRemoteAPI:    conf.ListenAddr,

